### PR TITLE
RSDK-6672: Object detection without category tensor

### PR DESCRIPTION
**Pull request by bhaney**
These edits allow the vision service to create an object detector without the requirement of having a categories tensor. 

This might be the case when a detector is only detecting one category, or the author decides to stuff the category information in another dimension. 

This also adds an optional parameter allowing the user to specify the bounding box order for their model

### CODE REVIEW

File: services/vision/mlvision/detector.go [hunk 1]
```diff
@@ -115,18 +120,42 @@ func attemptToBuildDetector(mlm mlmodel.Service, inNameMap, outNameMap *sync.Map
...
+			if err != nil {
+				return nil, err
+			}
+		}
+		// sometimes categories are stuffed into the score output. separate them out.
+		if !hasCategoryTensor {
+			shape := outMap[scoreName].Shape()
+			if len(shape) == 3 { // cartegories are stored in 3rd dimension
+				nCategories := shape[2]              // nCategories usually in 3rd dim, but sometimes in 2nd
+				if 4*nCategories == len(locations) { // it's actually in 2nd dim
```
Comment (kharijarrett) [inline services/vision/mlvision/detector.go:141]: Helpful comments. In general, I'm slightly worried about the criteria for knowing when the categories are just unreported and you need to look in the scores. We may have to come back to it but that feels like tomorrow's problem.

File: services/vision/mlvision/ml_model.go [hunk 2]
```diff
@@ -96,6 +97,27 @@ func registerMLModelVisionService(
...
+		for _, entry := range params.BoxOrder {
+			val, ok := checkOrder[entry]
+			if !ok || val { // if val is true, it means value was repeated
+				return nil, errors.Errorf(
+					"attribute xmin_ymin_xmax_ymax_order for model %q can only have entries 0, 1, 2 and 3, and only one instance of each. Got %v",
+					params.ModelName,
+					params.BoxOrder,
+				)
+			}
+			checkOrder[entry] = true
```
Comment (kharijarrett) [inline services/vision/mlvision/ml_model.go:118]: Okok I get it. I feel like there's an easier way to check for duplicates but you also had to write an argmax function soo...

File: services/vision/mlvision/detector.go [hunk 3]
```diff
@@ -187,7 +255,21 @@ func findDetectionTensorNames(outMap ml.Tensors, nameMap *sync.Map) (string, str
...
+	if okLoc && okScores { // names are known, just no categories
+		locString, ok := loc.(string)
+		if !ok {
+			return "", "", "", errors.Errorf("name map was not storing string, but a type %T", loc)
+		}
+		scoreString, ok := score.(string)
+		if !ok {
+			return "", "", "", errors.Errorf("name map was not storing string, but a type %T", score)
+		}
+		if len(outMap[scoreString].Shape()) == 3 || len(outMap) == 2 { // the categories are in the score
```
Comment (kharijarrett) [inline services/vision/mlvision/detector.go:267]: Feels weird but I don't know (I could be wrong)

File: services/vision/mlvision/ml_model_test.go [hunk 4]
```diff
@@ -278,6 +278,46 @@ func TestNewMLClassifier(t *testing.T) {
...
+	outNameMap.Store("score", "Identity_1")
+	gotDetector, err := attemptToBuildDetector(outModel, inNameMap, outNameMap, nil)
+	test.That(t, err, test.ShouldBeNil)
+	test.That(t, gotDetector, test.ShouldNotBeNil)
+
+	gotDetections, err := gotDetector(ctx, pic)
+	test.That(t, err, test.ShouldBeNil)
+	test.That(t, gotDetections[2297].Score(), test.ShouldBeGreaterThan, 0.7)
+	test.That(t, gotDetections[2297].Label(), test.ShouldResemble, "0")
+}
```
Comment (kharijarrett) [inline services/vision/mlvision/ml_model_test.go:319]: Could add a test for a model with actual (more than one) categories and then test to make sure we pull it out what they are from the scores. Basically test out extractCategoriesFromScores.

**Review by kharijarrett — APPROVED** [review]
LGTM.. A couple comments here and there but nothing close to dealbreaking.

File: services/vision/mlvision/ml_model_test.go (hunk 4 above)
Comment (bhaney) [reply services/vision/mlvision/ml_model_test.go:319]: good idea- doing this now

**Pull request merged**